File: Sorting Plateau/Divide and Conquer/Q2. Construct Binary Tree from Inorder and Postorder Traversal/Q2_ConstructBinaryTreefromInorderandPostorderTraversal.java
/**
 * Definition for a binary tree node.
 * public class TreeNode {
 *     int val;
 *     TreeNode left;
 *     TreeNode right;
 *     TreeNode() {}
 *     TreeNode(int val) { this.val = val; }
 *     TreeNode(int val, TreeNode left, TreeNode right) {
 *         this.val = val;
 *         this.left = left;
 *         this.right = right;
 *     }
 * }
 */
class Solution {
    private int postIndex; // Index pointer for postorder array
    private Map<Integer, Integer> inorderMap = new HashMap<>(); // value -> index in inorder

    public TreeNode buildTree(int[] inorder, int[] postorder) {
        postIndex = postorder.length - 1; // Start from last element of postorder

        // Build hashmap for inorder value -> index
        for (int i = 0; i < inorder.length; i++) {
            inorderMap.put(inorder[i], i);
        }

        // Build the tree recursively
        return helper(postorder, 0, inorder.length - 1);
    }

    private TreeNode helper(int[] postorder, int left, int right) {
        // Base case: no elements to construct tree
        if (left > right) {
            return null;
        }

        // Pick current root from postorder
        int rootVal = postorder[postIndex--];
        TreeNode root = new TreeNode(rootVal);

        // Root splits inorder into left and right subtrees
        int index = inorderMap.get(rootVal);

        // Important: build right subtree first because postorder goes left-right-root
        root.right = helper(postorder, index + 1, right);
        root.left = helper(postorder, left, index - 1);

        return root;
    }
}

File: Sorting Plateau/Divide and Conquer/Q2. Construct Binary Tree from Inorder and Postorder Traversal/Q2_ConstructBinaryTreefromInorderandPostorderTraversal.py
# Definition for a binary tree node.
# class TreeNode(object):
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right
class Solution(object):
    def buildTree(self, inorder, postorder):
        """
        :type inorder: List[int]
        :type postorder: List[int]
        :rtype: Optional[TreeNode]
        """
        # Map from value to index in inorder for quick lookup
        inorder_map = {val: idx for idx, val in enumerate(inorder)}
        # Index pointer for postorder (start from last element)
        self.post_index = len(postorder) - 1

        # Recursive helper function
        def helper(left, right):
            # If no elements to construct the tree
            if left > right:
                return None

            # Pick current root from postorder
            root_val = postorder[self.post_index]
            self.post_index -= 1
            root = TreeNode(root_val)

            # Find index of root in inorder
            index = inorder_map[root_val]

            # Important: build right subtree first
            root.right = helper(index + 1, right)
            root.left = helper(left, index - 1)

            return root

        return helper(0, len(inorder) - 1)
